runtime-light/stdlib/array/array-functions.h

```diff
@@ -5,16 +5,218 @@
 #pragma once
 
 #include <concepts>
+#include <cstdint>
 #include <functional>
 #include <utility>
 
 #include "runtime-common/core/runtime-core.h"
+#include "runtime-common/stdlib/array/array-functions.h"
 #include "runtime-light/coroutine/task.h"
 #include "runtime-light/stdlib/math/random-functions.h"
 #include "runtime-light/utils/concepts.h"
 
+namespace dl {
+
+template<typename T, typename Comparator>
+requires(std::invocable<Comparator, T, T>) task_t<void> sort(T *begin_init, T *end_init, Comparator compare) noexcept {
+  auto compare_call = [compare]<typename U>(U lhs, U rhs) -> task_t<int64_t> {
+    if constexpr (is_async_function_v<Comparator, U, U>) {
+      co_return co_await std::invoke(std::move(compare), std::move(lhs), std::move(rhs));
+    } else {
+      co_return std::invoke(std::move(compare), std::move(lhs), std::move(rhs));
+    }
+  };
+  T *begin_stack[32];
+  T *end_stack[32];
+
+  begin_stack[0] = begin_init;
+  end_stack[0] = end_init - 1;
+
+  for (int depth = 0; depth >= 0; --depth) {
+    T *begin = begin_stack[depth];
+    T *end = end_stack[depth];
+
+    while (begin < end) {
+      const auto offset = (end - begin) >> 1;
+      swap(*begin, begin[offset]);
+
+      T *i = begin + 1;
+      T *j = end;
+
+      while (true) {
+        while (i < j && (co_await compare_call(*begin, *i)) > 0) {
+          i++;
+        }
+
+        while (i <= j && (co_await compare_call(*j, *begin)) > 0) {
+          j--;
+        }
+
+        if (i >= j) {
+          break;
+        }
+
+        swap(*i++, *j--);
+      }
+
+      swap(*begin, *j);
+
+      if (j - begin <= end - j) {
+        if (j + 1 < end) {
+          begin_stack[depth] = j + 1;
+          end_stack[depth++] = end;
+        }
+        end = j - 1;
+      } else {
+        if (begin < j - 1) {
+          begin_stack[depth] = begin;
+          end_stack[depth++] = j - 1;
+        }
+        begin = j + 1;
+      }
+    }
+  }
+  co_return;
+}
+
+} // namespace dl
+
 namespace array_functions_impl_ {
 
+template<typename Result, typename U, typename Comparator>
+Result sort(array<U> &arr, Comparator comparator, bool renumber) noexcept {
```

Comment:
It's unsafe to pass an array to a coroutine by reference. In this function you access internal pointers of the array. Having such an array passed by reference, we can face with a case of the array being changed while we were blocked after `co_await` call.

Reply:
Sort functions works inplace

Reply:
make temporary copy